test/matchtests.jl

```diff
@@ -0,0 +1,334 @@
+    # Test single terms
+
+    myisodd(x::Int) = @match(x, i => i % 2 == 1)
+    @test filter(myisodd, 1:10) == filter(isodd, 1:10) == [1, 3, 5, 7, 9]
+end
+
+@testset "Alternatives, Guards" begin
+    # Alternatives, Guards
+
+    function parse_arg(arg::AbstractString, value::Any=nothing)
+        @match (arg, value) begin
+            ("-l",              lang),   if lang != nothing end => "Language set to $lang"
+            ("-o" || "--optim", n::Int),      if 0 < n <= 5 end => "Optimization level set to $n"
+            ("-o" || "--optim", n::Int)                         => "Illegal optimization level $(n)!"
+            ("-h" || "--help",  nothing)                        => "Help!"
+            bad                                                 => "Unknown argument: $bad"
+        end
+    end
+
+    @test parse_arg("-l", "eng")  == "Language set to eng"
+    @test parse_arg("-l")         == "Unknown argument: (\"-l\", nothing)"
+    @test parse_arg("-o", 4)      == "Optimization level set to 4"
+    @test parse_arg("--optim", 5) == "Optimization level set to 5"
+    @test parse_arg("-o", 0)      == "Illegal optimization level 0!"
+    @test parse_arg("-o", 1.0)    == "Unknown argument: (\"-o\", 1.0)"
+
+    @test parse_arg("-h") == parse_arg("--help") == "Help!"
+end
+
+# Regular Expressions
+
+# TODO: Fix me!
+
+# Note: the following test only works because Ipv4Addr and EmailAddr
+# are (module-level) globals!
+
+# const Ipv4Addr = r"(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})"
+# const EmailAddr = r"\b([A-Z0-9._%+-]+)@([A-Z0-9.-]+\.[A-Z]{2,4})\b"i
+
+# function regex_test(str)
+#     ## Defining these in the function doesn't work, because the macro
+#     ## (and related functions) don't have access to the local
+#     ## variables.
+
+#     # Ipv4Addr = r"(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})"
+#     # EmailAddr = r"\b([A-Z0-9._%+-]+)@([A-Z0-9.-]+\.[A-Z]{2,4})\b"i
+
+#     @match str begin
+#         Ipv4Addr(_, _, octet3, _),       if int(octet3) > 30 end => "IPv4 address with octet 3 > 30"
+#         Ipv4Addr()                                               => "IPv4 address"
+
+#         EmailAddr(_,domain), if endswith(domain, "ucla.edu") end => "UCLA email address"
+#         EmailAddr                                                => "Some email address"
+
+#         r"MCM.*"                                                 => "In the twentieth century..."
+
+#         _                                                        => "No match"
+#     end
+# end
+
+# @test regex_test("128.97.27.37")                 == "IPv4 address"
+# @test regex_test("96.17.70.24")                  == "IPv4 address with octet 3 > 30"
+
+# @test regex_test("beej@cs.ucla.edu")             == "UCLA email address"
+# @test regex_test("beej@uchicago.edu")            == "Some email address"
+
+# @test regex_test("MCMLXXII")                     == "In the twentieth century..."
+# @test regex_test("Open the pod bay doors, HAL.") == "No match"
```

Comment:
instead of commenting all of these out, maybe use `@test_broken` instead

Reply:
It wouldn't merely fail testing.  There would be a compile-time error in the @match that has invalid syntax (regular expression with parens and subpatterns).

Reply:
Also a pattern like `[a b]` would be a compile-time error.